src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -293,9 +293,19 @@ private static bool TopLevelChanged(SyntaxTree oldTree, SourceText oldText, Synt
         /// </summary>
         public bool HasContentChanged(DocumentState oldState)
         {
-            return oldState._treeSource != this._treeSource
-                || oldState.sourceTextOpt != this.sourceTextOpt
-                || oldState.textAndVersionSource != this.textAndVersionSource;
+            if (oldState.sourceTextOpt != this.sourceTextOpt
+                || oldState.textAndVersionSource != this.textAndVersionSource)
+            {
+                return true;
+            }
+            else if (oldState._treeSource == this._treeSource)
+            {
+                return false;
+            }
+            else
+            {
+                return oldState._treeSource.GetValue() == this._treeSource.GetValue();
```

Comment:
i'm confused... if these values are the same... then we return 'true' for 'HasContentChanged'?   i don't get how that's correct.  And if it is correct, we really need comments here explaining what is going on...

Reply:
Thanks for catching this :)